Quantized addition with 'out'

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24217 AddReLU module
* **#23971 Quantized addition with 'out'**

Differential Revision: [D16695592](https://our.internmc.facebook.com/intern/diff/D16695592)

Commit:
[2/2] Add out variant

Commit:
Update on "[2/2] Add out variant"


Differential Revision: [D16695592](https://our.internmc.facebook.com/intern/diff/D16695592)

Commit:
Update on "[2/2] Add out variant"


Differential Revision: [D16695592](https://our.internmc.facebook.com/intern/diff/D16695592)

jamesr66a commented on aten/src/ATen/native/quantized/cpu/qadd.cpp:
```diff
@@ -30,7 +31,7 @@ Tensor _add_out(Tensor& out, const Tensor& self, const Tensor& other) {
       const auto da = at::dequantize_val(self.q_scale(), self.q_zero_point(), a);
       const auto db = at::dequantize_val(other.q_scale(), other.q_zero_point(), b);
-      float c = da + db;
+      float c = float(da) + db;
```
Why is this cast needed? Doesn't dequantize_val already return a float?

 It's probably better to replace `auto` above with the correct types if that assumption is false

Commit:
Update on "[2/2] Add out variant"


Differential Revision: [D16695592](https://our.internmc.facebook.com/intern/diff/D16695592)

Commit:
Update on "Quantized addition with 'out'"


Differential Revision: [D16695592](https://our.internmc.facebook.com/intern/diff/D16695592)

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/qadd.cpp:
```diff
@@ -45,12 +46,23 @@ class QAdd final : public c10::OperatorKernel {
+                "Add operands must be the same size!");
+    TORCH_CHECK(qa.scalar_type() == qb.scalar_type(),
+                "Add operands should have same data type.");
+    return _add_out<ReLUFused>(out, qa, qb);
```
Please add a check for qscheme. We can mix and match per_tensor_affine and per_tensor_symmetric, but per_channel is not supported.

zafartahirov replied:
Resolved in the previous diff

Commit:
Update on "Quantized addition with 'out'"


Differential Revision: [D16695592](https://our.internmc.facebook.com/intern/diff/D16695592)

Commit:
Update on "Quantized addition with 'out'"


Differential Revision: [D16695592](https://our.internmc.facebook.com/intern/diff/D16695592)

Commit:
Update on "Quantized addition with 'out'"


Differential Revision: [D16695592](https://our.internmc.facebook.com/intern/diff/D16695592)
